engine/src/execution_graph/logic_controllers/BatchJoinProcessing.h

```diff
@@ -511,17 +572,123 @@ class JoinPartitionKernel :public kernel {
+					std::cout<<err<<std::endl;
+				}
+			}
+			ral::distribution::experimental::notifyLastTablePartitions(this->context.get(), ColumnDataMessage::MessageID());
+		});
+		
+		BlazingThread collect_small_table_thread([this, small_output_cache_name](){
+			auto  message_token = ColumnDataMessage::MessageID() + "_" + this->context->getContextCommunicationToken();
+			ExternalBatchColumnDataSequence external_input_left(this->context, message_token);
+			
+			while (external_input_left.wait_for_next()) {	
+				std::unique_ptr<ral::frame::BlazingHostTable> host_table = external_input_left.next();
+				this->add_to_output_cache(std::move(host_table), small_output_cache_name);
+			}
+		});
+
+		this->add_to_output_cache(std::move(big_table_batch), big_output_cache_name);
+
+		BlazingThread big_table_passthrough_thread([this, &big_table_sequence, big_output_cache_name](){
+			while (big_table_sequence.wait_for_next()) {	
+				auto batch = big_table_sequence.next();
+				this->add_to_output_cache(std::move(batch), big_output_cache_name);
+			}
+		});
+		
+		distribute_small_table_thread.join();
+		collect_small_table_thread.join();
+		big_table_passthrough_thread.join();		
+	}
+	
+	virtual kstatus run() {
+		
+		BatchSequence left_sequence(this->input_.get_cache("input_a"), this);
+		BatchSequence right_sequence(this->input_.get_cache("input_b"), this);
+
+		// lets parse part of the expression here, because we need the joinType before we load
+		std::string new_join_statement, filter_statement;
+		StringUtil::findAndReplaceAll(this->expression, "IS NOT DISTINCT FROM", "=");
+		split_inequality_join_into_join_and_filter(this->expression, new_join_statement, filter_statement);
+		std::string condition = get_named_expression(new_join_statement, "condition");
+		this->join_type = get_named_expression(new_join_statement, "joinType");
+
+		std::unique_ptr<ral::frame::BlazingTable> left_batch = left_sequence.next();
+		std::unique_ptr<ral::frame::BlazingTable> right_batch = right_sequence.next();
+
+		if (left_batch == nullptr || left_batch->num_columns() == 0){
+			std::cout<<"ERROR JoinPartitionKernel has empty left side and cannot determine join column indices"<<std::endl;
+		}
+
+		std::pair<bool, bool> scatter_left_right;
+		if (this->join_type == OUTER_JOIN){ // cant scatter a full outer join
+			scatter_left_right = std::make_pair(false, false);
+		} else {
+			scatter_left_right = determine_if_we_are_scattering_a_small_table(left_batch->toBlazingTableView(), 
+																				right_batch->toBlazingTableView());
+			if (scatter_left_right.first && this->join_type == LEFT_JOIN){
+				scatter_left_right.first = false; // cant scatter the left side for a left outer join
+			}
+		}
+		// scatter_left_right = std::make_pair(false, false); // Do this for debugging if you want to disable small table join optmization
+		if (scatter_left_right.first){
+			BatchSequenceBypass big_table_sequence(this->input_.get_cache("input_b"));
```

Comment:
Can you explain how bypass works?

Reply:
The BatchSequenceBypass allows you to take data from one cache and put it into another without calling decache. Which means it if was in system memory, it staying in system memory and not unnecessarily bringing it into GPU. 
In other words when you get something from a BatchSequenceBypass, you get a CacheData object and not a BlazingTable.